transports/grpc/server.go

```diff
@@ -0,0 +1,295 @@
+package grpctrans
+
+import (
+	"context"
+	"io"
+	"sync"
+
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+
+	"github.com/hashicorp/go-hclog"
+	"github.com/hashicorp/raft"
+	transportv1 "github.com/hashicorp/raft/proto/transport/v1"
+)
+
+var _ transportv1.RaftTransportServiceServer = &gRPCRaftService{}
+
+type gRPCRaftService struct {
+	rpcCh  chan raft.RPC
+	logger hclog.Logger
+
+	heartbeatFn     func(raft.RPC)
+	heartbeatFnLock sync.Mutex
+}
+
+func (g *gRPCRaftService) AppendEntries(ctx context.Context, req *transportv1.AppendEntriesRequest) (*transportv1.AppendEntriesResponse, error) {
+	var raftReq raft.AppendEntriesRequest
+	transportv1.AppendEntriesRequestToStruct(req, &raftReq)
+
+	respCh := make(chan raft.RPCResponse, 1)
+	rpc := raft.RPC{
+		Command:  &raftReq,
+		RespChan: respCh,
+	}
+
+	fastpathed := false
+	if isHeartbeat(req) {
+		g.heartbeatFnLock.Lock()
+		fn := g.heartbeatFn
+		g.heartbeatFnLock.Unlock()
+
+		if fn != nil {
+			// fastpath heartbeat processing
+			fn(rpc)
+			fastpathed = true
+		}
+	}
+
+	if !fastpathed {
+		select {
+		case g.rpcCh <- rpc:
+		case <-ctx.Done():
+			return nil, status.FromContextError(ctx.Err()).Err()
+		}
+	}
+
+	select {
+	case rpcResp := <-respCh:
+		if rpcResp.Error != nil {
+			return nil, status.Error(codes.Internal, rpcResp.Error.Error())
+		}
+
+		var resp transportv1.AppendEntriesResponse
+		transportv1.AppendEntriesResponseFromStruct(rpcResp.Response.(*raft.AppendEntriesResponse), &resp)
+		return &resp, nil
+	case <-ctx.Done():
+		return nil, status.FromContextError(ctx.Err()).Err()
+	}
+}
+
+func (g *gRPCRaftService) AppendEntriesPipeline(srv transportv1.RaftTransportService_AppendEntriesPipelineServer) error {
+	g.logger.Info("handling an append entries pipeline")
+	// This is not as efficient as it could be. To improve this we could
+	// queue up the append entries requests in one go routine and wait
+	// on their completion and send results back in a second go routine.
+	// If sending responses is experiencing extra latency then we would
+	// be better pipelining the requests. However the existing net_transport
+	// processes commands for a singular connection serially waiting on one
+	// RPC to finish before processing more commands. Therefore this should
+	// be at least as performant as that transport.
```

Comment:
The proposed optimization here is an alternative fix for #541 issue.

As implemented in the PR so far, this transport will have the same behavior as network transport that makes pipelining almost always worse than not pipelining at all.

If we did what this comment said though and applied batching on the recieve side here before writing to disk in one go, then that would fix the issue and we'd be able to avoid the problems of pipelining, at least in theory - we'd want to do actual performance comparisons to determine if there are other downsides.

I think ultimately it will end up being mostly equivalent though:
 - at low throughput pipelining doesn't impact perf at all since requests are too far apart that they complete before the next one arrives
 - at high throughput with sync receive processing, long pipelines prevent leader batching from grouping work which forces the pipeline to fill and the queue built up in the pipeline functions as a multiplier for latency.
 - at high throughput with receive side batching as proposed in this comment then you get back to the same latency as with pipelining disabled since the follower will automatically commit to disk more efficiently once a queue builds up and so keep the queue as small as possible (until we hit the max of 64 logs per batch or whatever limit we choose to prevent unbounded growth) it's not clear though if this actually is better than just not pipelining at all in the first place which acheives the same thing just by batching on the leader. In theory it might due to overlapping encoding and decoding and network RTT vs. purely sync replication but I _suspect_ if we test it we'd only get measurable gains where the RTT is a lot larger than the disk commit times which is not a typical situation for HashiCorp raft usages at least.

This is more of a brain dump for when this is picked up than a recommendation at this point but we should do at least one of:
 - make pipeline size configurable and default to short like in #541 
 - _or_ batch on the receive side here too as in the comment.

Reply:
I think after all your investigation we have generally proved that pipelining in the transport on the sending side is almost always worse in cases where this library is intended to be used.

More or less we want to cause the followers to commit to disk fewer times with larger batches when under heavier load. It doesn't seem worth it to batch things prior to hitting the transport on the leader and then again on the follower. 
So given that we already have batching code outside of the transport on the leader, it seems like we ought to keep this transport simpler and not do receive side batching.